Add auto start to particle emitters

gamecoder-nz opened the pull request:
This fixes the issue where if the Emitting flag is true and the One shot flag is true then Emitting flag is set to false when the emitter has finished and it won't be started if instantiated into the game.  Setting this flag to true will ensure that the emitter is started when added to the game.

* *Bugsquad edit, closes: https://github.com/godotengine/godot/issues/97213*

gamecoder-nz committed:
Fixed one shot emitting issue

If one shot is enabled then emitting will still be left to true.  This will ensure that the particle emitter will start emitting when it is placed into the game

gamecoder-nz committed:
Added auto start feature to emitters.  It will start the emitter once it is placed in the scene.

gamecoder-nz committed:
Change auto_start to autostart to make it consistent with timer

gamecoder-nz committed:
Fixed invalid character issue

KoBeWi commented:
I compared it to Timer's implementation and there are 2 differences:
- Particles will start emitting on ENTER_TREE, which means they will emit every time they are added to scene tree. This is unlike timer, which uses READY and starts only once. Although maybe for particles the expectation is different and it makes sense, idk.
- Auto Start takes effect in editor, so particles emit every time you open the scene. This should be fixed.

Also, since particles have `emitting` property, setting autostart does not make sense when particles are already emitting, and possibly when they aren't one-shot (because why would you disable non-one-shot emitting and use autostart instead?). The property should be disabled accordingly, using `_validate_property()`.

gamecoder-nz committed:
Moved autostart check to ready notification
Autostart does not activate in editor
Autostart property does not display in editor if emitting is true and oneshot is false

gamecoder-nz commented:
I have made the necessary changes.  For the _validate_property I notice that it will only do this when the scene is opened.  Is there a way to disable the property while the scene is open?

KoBeWi commented:
> Is there a way to disable the property while the scene is open?

`notify_property_list_changed()`

KoBeWi commented on scene/2d/cpu_particles_2d.cpp:
```diff
@@ -532,6 +540,13 @@ void CPUParticles2D::_validate_property(PropertyInfo &p_property) const {
+
+	if (p_property.name == "autostart") {
+		if (emitting && !one_shot) {
+			p_property.usage = PROPERTY_USAGE_NONE;
```
PROPERTY_USAGE_READ_ONLY would be better.

QbieShay commented:
Hey 👋 

I understand why this PR is opened, but I'd like to eventually redesign the way particle emission works. Oneshot currently has LOTS of problems.

I would prefer to come up wtih a solution that addresses the base issue (oneshot being weird) rather than addressing the symptom.  This autostart property is very simple to otherwise implement in gdscript, I don't think there's an immediate need to add this to the engine.

Nevertheless thank you for taking interest in contributing on particles! If you want to discuss more the approaches to emission, we can chat on [discord](https://discord.gg/HX4xAGaGjm)

gamecoder-nz committed:
Updated usage to  PROPERTY_USAGE_EDITOR | PROPERTY_USAGE_READ_ONLY
Changing emitting or one shot disables autostart

QbieShay commented:
Thank you for your pull request, but I do not plan to merge this code.

I mentioned it in related proposals, but i want the way emission works to be improved with a significant rework.
(see https://github.com/godotengine/godot-proposals/issues/7322#issuecomment-1780618682)

Since what your PR does can be done in a small gdscript file, I'll close the PR.

Nevertheless, thank you for your effort in improving Godot!
